crates/iceberg/src/spec/schema.rs

```diff
@@ -642,6 +644,204 @@ impl SchemaVisitor for IndexByName {
     }
 }
 
+struct PruneColumn {
+    selected: HashSet<i32>,
+    select_full_types: bool,
+}
+
+/// Visit a schema and returns only the fields selected by id set
+pub fn prune_columns(
+    schema: &Schema,
+    selected: HashSet<i32>,
+    select_full_types: bool,
+) -> Result<Option<Type>> {
+    let mut visitor = PruneColumn::new(selected, select_full_types);
+    visit_schema(schema, &mut visitor)
+}
+
+impl PruneColumn {
+    fn new(selected: HashSet<i32>, select_full_types: bool) -> Self {
+        Self {
+            selected,
+            select_full_types,
+        }
+    }
+
+    fn project_selected_struct(projected_field: Option<Type>) -> Result<StructType> {
+        match projected_field {
+            // If the field is a StructType, return it as such
+            Some(Type::Struct(s)) => Ok(s),
+            Some(_) => Err(Error::new(
+                ErrorKind::DataInvalid,
+                "Projected Field must be Struct".to_string(),
+            )),
+            // If projected_field is None or not a StructType, return an empty StructType
+            None => Ok(StructType::default()),
+        }
+    }
+    fn project_list(list: &ListType, element_result: Type) -> Result<ListType> {
+        if *list.element_field.field_type == element_result {
+            return Ok(list.clone());
+        }
+        Ok(ListType {
+            element_field: Arc::new(NestedField {
+                id: list.element_field.id,
+                name: list.element_field.name.clone(),
+                required: list.element_field.required,
+                field_type: Box::new(element_result),
+                doc: list.element_field.doc.clone(),
+                initial_default: list.element_field.initial_default.clone(),
+                write_default: list.element_field.write_default.clone(),
+            }),
+        })
+    }
+    fn project_map(map: &MapType, value_result: Type) -> Result<MapType> {
+        if *map.value_field.field_type == value_result {
+            return Ok(map.clone());
+        }
+        Ok(MapType {
+            key_field: map.key_field.clone(),
+            value_field: Arc::new(NestedField {
+                id: map.value_field.id,
+                name: map.value_field.name.clone(),
+                required: map.value_field.required,
+                field_type: Box::new(value_result),
+                doc: map.value_field.doc.clone(),
+                initial_default: map.value_field.initial_default.clone(),
+                write_default: map.value_field.write_default.clone(),
+            }),
+        })
+    }
+}
+
+impl SchemaVisitor for PruneColumn {
+    type T = Option<Type>;
+
+    fn schema(&mut self, _schema: &Schema, value: Option<Type>) -> Result<Self::T> {
+        Ok(Some(value.unwrap()))
+    }
+
+    fn field(&mut self, field: &NestedFieldRef, value: Self::T) -> Result<Self::T> {
```

Comment:
```suggestion
    fn field(&mut self, field: &NestedFieldRef, value: Option<Type>) -> Result<Self::T> {
```

Sorry for not being clear, I mean we should replace all `Self::T` with concrete type for better readability.